Added explanation for silver milk visits and CPP socdist Solution (#744)
cpinitiative/usaco-guide: solutions/USACO dec19-milkvisits-silver.mdx

id: usaco-968
source: USACO Silver 2019 December
title: Milk Visits
author: Qi Wang, Tanish Tyagi
---

[Official Analysis](http://www.usaco.org/current/data/sol_milkvisits_silver_dec19.html)

**Time Complexity:** $\mathcal{O}(N + M)$

We can use a DFS to create connected components from an adjacency list (which is made when we process the roads between the farms). Each connected component will contain farms that all have the same breed. 

> Graph based off sample test case: <br/>

Reviewer: ```suggestion
Adjacency list of the sample graph: <br/>
```

Reviewer: Also you should include a photo/diagram of the sample graph.

> 1 (H) --> 2 (H), 5 (G) <br/>

Reviewer 2: I'd just remove the breed completly, it looks sketchy rendered

> 2 (H) --> 3, 4 (H) <br/>

Reviewer: 2 also points to 1?

3 (G) --> 2 (H) <br/>
4 (H) --> 2 (H) <br/>
5 (G) --> 1 (G) <br/>

> NOTE: We are not going to store the breed of cow in our adjacency list. The breed is there for convenience.

Reviewer 2: > The breed is there for convenience.

This may confuse people, so just delete it?

Connected Components based off of graph: <br/>

> components (first index is 1) = [1, 1, 2, 1, 3] <br/> <br/>

Reviewer 2: One indexed instead of first index is 1

> components (first index is 1) = [1, 1, 2, 1, 3] <br/> <br/>

Reviewer 2: try latex with variable names

> Farms 1,2, and 4 are all Holsteins and make up component 1, farm 3 makes up component 2, and farm 5 makes up component 3. 

Reviewer: Explanation isn't super intuitive. Maybe try something like this?:

```
There are three connected components of the same cow breed in this graph:

- Component 1: Farms 1, 2, and 4
- Component 2: Farm 3
- Component 3: Farm 5

Create an array $\texttt{comp}$ such that $\texttt{comp}[A]$ is the component of farm $A$.

$\texttt{comp} = [1, 1, 2, 1, 3]$ (1-indexed)
```

> When it comes to seeing which farmers will be happy, we just have to evalulate 3 conditions: <br/>

Reviewer 2: We evaluate 3 conditions to check whether the farmer will be happy:

> 1. Farm $A$ and $B$ are part of the same component. If $C$ is the same breed as $A$ / $B$, then we should output 1.

Reviewer: ```suggestion
1. Farm $A$ and $B$ are part of the same component.  This means the path between $A$ and $B$ only contains cows of the same breed. If the farmer prefers milk from this breed, then we should output 1.
```

id: usaco-968
source: USACO Silver 2019 December
title: Milk Visits
author: Qi Wang, Tanish Tyagi
---

[Official Analysis](http://www.usaco.org/current/data/sol_milkvisits_silver_dec19.html)

**Time Complexity:** $\mathcal{O}(N + M)$

We can use a DFS to create connected components from an adjacency list (which is made when we process the roads between the farms). Each connected component will contain farms that all have the same breed. 

Adjacency List based off sample test case: <br/>
$1$ $\rightarrow$ $2$, $5$  <br/>
$2$ $\rightarrow$ $1$, $3$, $4$  <br/>
$3$ $\rightarrow$ $2$ <br/>
$4$ $\rightarrow$ $2$ <br/>

> $5$ $\rightarrow$ $1$ <br/>

Reviewer 3: ```
Adjacency List based off sample test case:

$1\rightarrow 2,5$

$2\rightarrow 1,3,4$

$3\rightarrow 2$

$4\rightarrow 2$

$5\rightarrow 1$
```
is possibly neater

Reviewer: ^ yes do this please

> Adjacency List based off sample test case: <br/>
> $1$ $\rightarrow$ $2$, $5$  <br/>
> $2$ $\rightarrow$ $1$, $3$, $4$  <br/>
> $3$ $\rightarrow$ $2$ <br/>
> $4$ $\rightarrow$ $2$ <br/>
> $5$ $\rightarrow$ $1$ <br/>

Reviewer: ```suggestion
Adjacency List based off sample test case:

$1\rightarrow 2,5$

$2\rightarrow 1,3,4$

$3\rightarrow 2$

$4\rightarrow 2$

$5\rightarrow 1$
```

There are three connected components of the same cow breed in this graph:
- Component $1$: Farms $1$, $2$, and $4$
- Component $2$: Farm $3$
- Component $3$: Farm $5$

Create an array $\texttt{comp}$ such that $\texttt{comp}[A]$ is the component of farm $A$.

$\texttt{comp} = [1, 1, 2, 1, 3]$ (1-indexed)

We evaluate 3 conditions to check whether the farmer will be happy:

> 1. Farm $A$ and $B$ are part of the same component.  This means the path between $A$ and $B$ only contains cows of the same breed. If the farmer prefers milk from this breed, then we should output $1$.

Reviewer: ```suggestion
1. Farms $A$ and $B$ are part of the same component.  This means the path between $A$ and $B$ only contains cows of the same breed. If the farmer prefers milk from this breed, then we should output $1$.
```

> 2. Farm $A$ and $B$ are part of the different components. This means that the farmer will always be satisfied because the path between $A$ and $B$ contains both breeds of cows. We should output $1$.

Reviewer: ```suggestion
2. Farms $A$ and $B$ are part of different components. This means that the farmer will always be satisfied because the path between $A$ and $B$ contains both breeds of cows. We should output $1$.
```

> 3. If case $1$ and $2$ are not met, then we should output 0.

Reviewer: ```suggestion
3. If cases $1$ and $2$ are not met, then we should output 0.
```

Queries for Sample Test Case: 
1. Farms $1$ and $4$ are in the same component, and since Farmer $1$'s preferred cow is Holstein, they will be satisfied. ($1$)
2. Farms $1$ and $4$ are in the same component, but since Farmer $2$'s preferred cow is Guernsey, they will be unsatisfied. ($0$)

> 3. Farms $1$ and $3$ are in different components, so Farmer $3$ will be satisfied. (1)

Reviewer: ```suggestion
3. Farms $1$ and $3$ are in different components, so Farmer $3$ will be satisfied. ($1$)
```

to stay consistent with the other queries

id: usaco-968
source: USACO Silver 2019 December
title: Milk Visits
author: Qi Wang, Tanish Tyagi
---

[Official Analysis](http://www.usaco.org/current/data/sol_milkvisits_silver_dec19.html)

**Time Complexity:** $\mathcal{O}(N + M)$

We can use a DFS to create connected components from an adjacency list (which is made when we process the roads between the farms). Each connected component will contain farms that all have the same breed. 

Adjacency List based off sample test case: <br/>

> 1 --> 2, 5  <br/>

Reviewer 2: can u add latex to this section? personally i think it looks better with latex.

\rightarrow for the arrow btw.

2 --> 1, 3, 4  <br/>
3 --> 2 <br/>
4 --> 2 <br/>
5 --> 1 <br/>

> ![](<USACO-dec19-milk-visits-silver/graph.png>)

Reviewer 2: can u make this picture more formal lol?
